foundation/org.eclipse.persistence.core/src/main/java/org/eclipse/persistence/internal/helper/ConcurrencyManager.java

```diff
@@ -84,14 +85,39 @@ public void acquire() throws ConcurrencyException {
      * called with true from the merge process, if true then the refresh will not refresh the object
      */
     public synchronized void acquire(boolean forMerge) throws ConcurrencyException {
+
+        //FIX - BUG-30604379 - Flag the time when we start the while loop
+        final Date whileStartDate = new Date();
+        Thread currentThread = Thread.currentThread();
+        DeferredLockManager lockManager = getDeferredLockManager(currentThread);
+
         while (((this.activeThread != null) || (this.numberOfReaders > 0)) && (this.activeThread != Thread.currentThread())) {
             // This must be in a while as multiple threads may be released, or another thread may rush the acquire after one is released.
             try {
                 this.numberOfWritersWaiting++;
-                wait();
-                this.numberOfWritersWaiting--;
+
+                // FIX - BUG-30604379 - Do not wait forever - max allowed time to wait is 10 second and then check conditions again
+                // wait();
+                wait(10000l);
```

Comment:
This value should be configurable with a property instead of a random value. A default value of `0` will preserve the original `wait()` call.